Private token support for GitHub and npm URLs

mariusandra opened the pull request:
## Changes

- In addition to Github, now private token support is also enabled for GitHub and npm URLs. 
- This is necessary to make plugin updates work with github, where we (or the project) will need to enter a global API token in order to not be hit with the 60 requests/min API limit.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

mariusandra committed:
private token support for github and npm

Twixes commented on frontend/src/scenes/plugins/install/CustomPlugin.tsx:
```diff
@@ -13,7 +13,8 @@ export function CustomPlugin(): JSX.Element {
                 <p>
-                    To install a third-party or custom plugin, please paste the plugin's repository below.
+                    To install a third-party or custom plugin, paste its GitHub, GitLab or npm URL below. For private
+                    repositories, append <code>?private_token=TOKEN</code> at the end of the URL.
```
This should link to what the `TOKEN` should be (i.e.: [GitHub](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token), [GitLab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html), [npm](https://docs.npmjs.com/creating-and-viewing-access-tokens)).

Twixes commented on posthog/plugins/utils.py:
```diff
@@ -11,6 +11,12 @@
 
 
 def parse_github_url(url: str, get_latest_if_none=False) -> Optional[Dict[str, Optional[str]]]:
+    private_token = None
+    if "?" in url:
+        url, query = url.split("?")
+        params = {k: v[0] for k, v in parse_qs(query).items()}
+        private_token = params.get("private_token", None)
+
```
Since we're using this logic in a few places now, may be cleaner extracted to its own function.

mariusandra committed:
refactor url splitting

mariusandra committed:
mypy

mariusandra committed:
fix needless regexp escapes

mariusandra committed:
simplify parse_url output, tag is always tag now

mariusandra committed:
simplify parsing api response to get latest tag

mariusandra committed:
add links to api key docs for various providers

mariusandra committed:
update title

mariusandra commented:
Thanks! Did some slight improvements to the code and the interface:

![image](https://user-images.githubusercontent.com/53387/105704419-10723380-5f0f-11eb-9020-0ba128ab084a.png)

mariusandra committed:
global _TOKEN settings for github, gitlab and npm (mainly to bump api limits)

mariusandra committed:
move gitlab api key to headers

mariusandra commented:
I added two other things and it's ready for a review again. 

- GitLab API requests now also use a Header instead of passing the token in the URL query string
- I added global `NPM_TOKEN`, `GITLAB_TOKEN` and `GITHUB_TOKEN` settings. The idea is, at least for cloud, to use a custom GITHUB token in order to get around the 60 requests/hour limit, which makes plugins unusable after a while. Every time we check for a new version or install a plugin, this API is hit, and the 1 requests per minute limit is way too low. 
- This could be set per Project/Organization later, but that's scope creep right now.

Twixes reviewed (approved):
Looks good.
Didn't notice any way to extract the tokens out of PostHog. 🤔

mariusandra commented:
I think that's a good thing...? 🤔  See also #3082

Thanks for the review! Will fix the conflict and merge.

mariusandra committed:
Merge branch 'master' into plugin-api-keys
